### DISCUSSION TITLE
ArrayInstance class not opened for extenstions

### DISCUSSION THREAD

**Discussion by xtadex** [sebastienros/jint#1909] [General]
While migrating from V2 to V3, I noticed that in V3, the ArrayInstance is not opened for extension (all virtual methods are marked as sealed). I used a virtual method for tracking changes in the array by overriding these methods.

**Comment by lahma** [comment]
Can you share some code which would show the requirements and use case for you. I'd think Proxy could be an answer to this but I would need more details.

**Reply by xtadex** [reply]
Hi @lahma , please find the link below to the class inherited fromArrayInstance
https://github.com/TSSArt/Xtate.DataModel.EcmaScript/blob/main/src/Xtate.DataModel.EcmaScript/DataModelArrayWrapper.cs

**Reply by lahma** [reply]
What will be missing if you remove your own wrapper and let Jint handle the wrapping? Jint has come a long way and can expose IList types quite well.

**Reply by xtadex** [reply]
My framework keeps objects(key/value pairs) and arrays. I also need to allow JS code to modify these objects. In V2, I did it through wrappers: DataModelObjectWrapper and DataModelArrayWrapper. 

Both classes are responsible for the instant propagation of changes in a JS object or array to a DataModel object or array.

It worked like a charm in V2.
In V3, I can extend ObjectInstance, but ArrayInstance has all virtual methods sealed.

**Reply by xtadex** [reply]
Potentially, I can create an entirely new CustomArrayInstance derived from ObjectInstance, but it will look a bit ugly from an architectural point of view

**Reply by lahma** [reply]
It's just that your implementation looks somewhat redundant now that Jint can expose types like IList in quite logical way. So example code without your custom wrapper and Jint wrapping and showing missing bits would be great.

I'm hoping Jint could cater these interop scenarios without much custom code.

**Reply by xtadex** [reply]
If Jint can wrap my custom IList, it will be perfect. Then, I can intercept all array updates inside the IList implementation. 

Do you think it will work for me?

`
IList mylist = new DataModelList();
var jsArray = JsValue.FromObject(engine, my);
`

If the JS script modifies jsArray, I will see all changes immediately in the mylist object. Correct?

**Reply by lahma** [reply]
This should be the case, but please verify.

**Reply by xtadex** [reply]
I've checked this approach. 
The quick test console app is here https://github.com/xtadex/JintV3Test/blob/main/JintV3Test/Program.cs

It can read/write elements inside of IList<>, but it fails on
- write outside of upper bound (in JS array length adjusted to make write operation successful)
- the length of the underlying array can't be changed by any of operations:
  - set length property
  - calling push()
  - write above upper boundary

**Reply by lahma** [reply]
Would you like to take a stab to improve the current interop logic? I'm not sure which parts might need a separate interop configuration switch so they wouldn't be surprising.
